### ISSUE TITLE
Custom predict function for keras

### ISSUE THREAD

**Issue by snvv** [tripartio/ale#10]
Hello,

Does the `ale` package work with models that use embeddings? I'm using the following deep learning model in R:

```r
library(tensorflow)
library(keras)
library(ale)
library(tidyverse)

set.seed(123)
n <- 1000
df <- tibble(
  firm_id = sample(1:50, n, replace = TRUE),  # Categorical variable requiring embeddings
  sector = sample(1:10, n, replace = TRUE),   # Another categorical variable
  x1 = rnorm(n),  # Continuous variable
  x2 = rnorm(n),  
  y = 3*x1 + 2*x2 + firm_id*0.1 + sector*0.2 + rnorm(n)  # Target variable
)

# Convert categorical variables to factors
df$firm_id <- as.factor(df$firm_id)
df$sector <- as.factor(df$sector)

# One-hot encode categorical variables (Not actually one-hot encoding, but converting to integers for embeddings)
df <- df %>%
  mutate(firm_id = as.integer(firm_id) - 1,  # Convert factor to integer (0-based)
         sector = as.integer(sector) - 1)

# Split into training and test sets
train_idx <- sample(1:n, size = 0.8 * n)
train_data <- df[train_idx, ]
test_data <- df[-train_idx, ]


# Define input layers
firm_input <- layer_input(shape = 1, dtype = "int32", name = "firm_input")
sector_input <- layer_input(shape = 1, dtype = "int32", name = "sector_input")
x1_input <- layer_input(shape = 1, name = "x1_input")
x2_input <- layer_input(shape = 1, name = "x2_input")

# Embedding layers
firm_embedding <- firm_input %>%
  layer_embedding(input_dim = 50, output_dim = 3) %>%
  layer_flatten()

sector_embedding <- sector_input %>%
  layer_embedding(input_dim = 10, output_dim = 2) %>%
  layer_flatten()

# Concatenate all inputs
concatenated <- layer_concatenate(list(firm_embedding, sector_embedding, x1_input, x2_input))

# Fully connected layers
dense_1 <- concatenated %>%
  layer_dense(units = 10, activation = "relu")

output <- dense_1 %>%
  layer_dense(units = 1)

# Build model
model <- keras_model(inputs = list(firm_input, sector_input, x1_input, x2_input), outputs = output)

# Compile model
model %>% compile(
  optimizer = "adam",
  loss = "mse"
)

# Convert data into lists (TensorFlow requires list inputs for multi-input models)
x_train <- list(train_data$firm_id, train_data$sector, train_data$x1, train_data$x2)
y_train <- train_data$y

x_test <- list(test_data$firm_id, test_data$sector, test_data$x1, test_data$x2)
y_test <- test_data$y

# Train the model
history <- model %>%
  fit(
    x = x_train,
    y = y_train,
    epochs = 100,
    batch_size = 32,
    validation_data = list(x_test, y_test),
    verbose = 1
  )

# Make predictions on test data (This line is not needed for ALE, but kept for reference)
# test_data <- model %>% predict(x_test)


predict_function_ale <- function(model, newdata, type = "response") {
  # Convert categorical variables (crucial for ALE to work correctly)
  newdata <- newdata %>%
    mutate(firm_id = as.integer(firm_id) - 1,
           sector = as.integer(sector) - 1)

  x_input <- list(newdata$firm_id, newdata$sector, newdata$x1, newdata$x2)
  preds <- model %>% predict(x_input)

  return(as.numeric(preds))
}

ale_results <- ale(
  data = test_data,  # Use the original test data frame, not the predictions
  pred_fun = predict_function_ale,
  pred_type = "response",
  y_col = "y"
)

print(ale_results)

Best Regards

**Comment by tripartio** [comment]
I had to spend quite a bit of time getting this to work, not so much because of the `ale` package but because it's quite a pain to work with Python-backend packages in R like `tensorflow` and `keras`. In the end, the challenge has nothing to do with the fact that model is deep learning or based on embeddings; it's just a matter of the idiosyncracies of `keras`.  (On a side note, have you ever tried [torch for R](https://torch.mlverse.org/)? It's a native R implementation that works just like PyTorch. No Python needed.)

The main issue is that the `keras` `predict()` function requires a list as the data input rather than a dataframe. But the `ale` package requires a dataframe as input to `ale()` because, not only is that the R standard, but it needs to work with the column names. So, here's the corrected custom input function and `ale()` call that works (hopefully, the comments are self:

```r
# Make predictions on test data.
# This step is always necessary towards creating a custom predict function.
test_preds <- predict(model, x_test)


# Custom predict function for keras model objects.
# Note that the ale package requires EXACTLY the following argument names for a custom
# predict function. These are R standards; they were not invented by the ale package.
keras_pred_fun <- function(
    object,  # you cannot change this to "model" or anything else
    newdata, 
    # Since type isn't used, just set it to NULL. But it must still be included as an argument.
    type = NULL  
) {
  predict(
    object = object, 
    # The main challenge with keras is that it inputs a list for predictions instead of a dataset. 
    # So, the custom predict function must convert the dataframe required by ale() 
    # into the list format required by keras.
    x = newdata |>  # the keras predict argument is called x
      select(-any_of('y')) |>  # you must remove the y (outcome column from the dataframe)
      as.list() |>  # the keras x argument must be a list, not a dataframe
      unname(),  # the dataframe column names must be deleted or else keras predict errors
      verbose = 0  # silence the unnecessary messages during keras prediction
  )
}

ale_results <- ale(
  test_data,
  model,
  y_col = 'y',  # required for non-standard R models like keras
  pred_fun = keras_pred_fun,
  # parallel = 0
)

# Print the resulting plots
ale_results$plots$firm_id
ale_results$plots$sector
ale_results$plots$x1
ale_results$plots$x2
```

![Image](https://github.com/user-attachments/assets/4076c771-84e6-4968-a4c4-28ee386e5dc9)

![Image](https://github.com/user-attachments/assets/731924a1-bfbb-4a22-938c-c3341e769855)

![Image](https://github.com/user-attachments/assets/f78979f6-f04d-48ee-9f23-7d90b4ad32b2)

![Image](https://github.com/user-attachments/assets/b1cca07d-f3db-42ff-a589-46a7adfb079a)

**Comment by snvv** [comment]
Thank you so much for your help with this. I really appreciate you taking the time and effort to find such an effective solution.  It's going to make a big difference.
Best Regards

**Comment by tripartio** [comment]
@snvv I'd like to give you a heads-up, though: I'm in the process of completely rewriting the ale package. The development version already has most of the new functionality: categorical y values, S7 objects, and more. Unfortunately, this will come at the cost of a new functional interface but I think the added functionality will be worthwhile. I hope to publish the new version to CRAN by the end of this month.

If you have any burning wishes for the package, please create a new issue for each feature suggestion.

**Comment by snvv** [comment]
I tested the functionality of your proposed keras_pred_fun, and encountered an error during the calculation of ale. Here are the details:

#### Progress Output Before Interruption
```
25% | Calculating ALE ETA: 35s
Progress interrupted by purrr_error_indexed condition: ℹ In index: 1.
Caused by error in `map()`:
ℹ In index: 1.
Caused by error in `do.call()`:
! 'what' must be a function or character string
Error in (function (.x, .f, ..., .progress = FALSE)  : ℹ In index: 1.
Caused by error in `map()`:
ℹ In index: 1.
Caused by error in `do.call()`:
! 'what' must be a function or character string
```

#### Custom Prediction Function

I attempted to create a custom prediction function to address the issue:

```r
keras_pred_fun <- function(
  object,  # Keras model object
  newdata,  # Data frame containing input data
  type = NULL  # Unused, but required
) {
  # Split the input data frame into separate tensors
  firm_input <- as.matrix(newdata$firm_id)
  sector_input <- as.matrix(newdata$sector)
  x1_input <- as.matrix(newdata$x1)
  x2_input <- as.matrix(newdata$x2)
  
  # Make predictions using the 4 separate inputs
  predictions <- predict(
    object = object,
    x = list(firm_input, sector_input, x1_input, x2_input),  # List of inputs
    verbose = 0
  )
  
  # Return predictions as a numeric vector
  return(as.numeric(predictions))
}
```

However, this also resulted in the same error:

```
25% | Calculating ALE ETA: 34s
Progress interrupted by purrr_error_indexed condition: ℹ In index: 1.
Caused by error in `map()`:
ℹ In index: 1.
Caused by error in `do.call()`:
! 'what' must be a function or character string
Error in (function (.x, .f, ..., .progress = FALSE)  : ℹ In index: 1.
Caused by error in `map()`:
ℹ In index: 1.
Caused by error in `do.call()`:
! 'what' must be a function or character string
```

Despite these errors, the custom prediction function `keras_pred_fun` appears to work correctly when debugging:

```r
[1] "Predictions (numeric vector):"
[1] 1.97840726  7.48064804  8.63505840  2.40812063  5.32712603
```

### Minimal Reproducible Example

Here is a minimal reproducible example of the code and data:

```r
library(tensorflow)
library(keras)
library(ale)
library(tidyverse)

# Simulate data
set.seed(123)
n <- 1000
df <- tibble(
  firm_id = sample(1:50, n, replace = TRUE),
  sector = sample(1:10, n, replace = TRUE),
  x1 = rnorm(n),
  x2 = rnorm(n),
  y = 3 * x1 + 2 * x2 + firm_id * 0.1 + sector * 0.2 + rnorm(n)
)

# Convert categorical variables to integers
df <- df %>%
  mutate(firm_id = as.integer(firm_id) - 1,
         sector = as.integer(sector) - 1)

# Split into training and test sets
train_idx <- sample(1:n, size = 0.8 * n)
train_data <- df[train_idx, ]
test_data <- df[-train_idx, ]

# Define and train the Keras model
firm_input <- layer_input(shape = 1, dtype = "int32", name = "firm_input")
sector_input <- layer_input(shape = 1, dtype = "int32", name = "sector_input")
x1_input <- layer_input(shape = 1, name = "x1_input")
x2_input <- layer_input(shape = 1, name = "x2_input")

firm_embedding <- firm_input %>%
  layer_embedding(input_dim = 50, output_dim = 3) %>%
  layer_flatten()

sector_embedding <- sector_input %>%
  layer_embedding(input_dim = 10, output_dim = 2) %>%
  layer_flatten()

concatenated <- layer_concatenate(list(firm_embedding, sector_embedding, x1_input, x2_input))

dense_1 <- concatenated %>%
  layer_dense(units = 10, activation = "relu")

output <- dense_1 %>%
  layer_dense(units = 1)

model <- keras_model(inputs = list(firm_input, sector_input, x1_input, x2_input), outputs = output)

model %>% compile(
  optimizer = "adam",
  loss = "mse"
)

history <- model %>%
  fit(
    x = list(train_data$firm_id, train_data$sector, train_data$x1, train_data$x2),
    y = train_data$y,
    epochs = 10,
    batch_size = 32,
    validation_data = list(
      list(test_data$firm_id, test_data$sector, test_data$x1, test_data$x2),
      test_data$y
    ),
    verbose = 0
  )

# Custom prediction function
keras_pred_fun <- function(
  object,  # Keras model object
  newdata,  # Data frame containing input data
  type = NULL  # Unused, but required
) {
  # Split the input data frame into separate tensors
  firm_input <- as.matrix(newdata$firm_id)
  sector_input <- as.matrix(newdata$sector)
  x1_input <- as.matrix(newdata$x1)
  x2_input <- as.matrix(newdata$x2)
  
  # Make predictions using the 4 separate inputs
  predictions <- predict(
    object = object,
    x = list(firm_input, sector_input, x1_input, x2_input),  # List of inputs
    verbose = 0
  )
  
  # Return predictions as a numeric vector
  return(as.numeric(predictions))
}

# Calculate ALE
ale_results <- ale(
  test_data,  # Input data (must be a data frame)
  model,      # Your Keras model
  y_col = 'y',  # Target column
  pred_fun = keras_pred_fun  # Custom prediction function
)
```

### Questions

1. **Why does the ALE package fail with the `do.call()` error when using the custom prediction function?**
2. **Is there a specific format or structure that the custom prediction function must adhere to for multi-input models?**
3. **Are there any known issues with the ALE package and Keras models?**

Thank you in advance for your help.

**Comment by tripartio** [comment]
After running the code I gave you (successfully again without any errors), here is my session information:

```r
> tensorflow::tf_version()
[1] ‘2.16’
> sessionInfo()
R version 4.4.2 (2024-10-31 ucrt)
Platform: x86_64-w64-mingw32/x64
Running under: Windows 11 x64 (build 26100)

Matrix products: default


locale:
[1] LC_COLLATE=French_France.utf8  LC_CTYPE=French_France.utf8    LC_MONETARY=French_France.utf8
[4] LC_NUMERIC=C                   LC_TIME=French_France.utf8    

time zone: Europe/Paris
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] lubridate_1.9.4   forcats_1.0.0     stringr_1.5.1     dplyr_1.1.4       purrr_1.0.2      
 [6] readr_2.1.5       tidyr_1.3.1       tibble_3.2.1      ggplot2_3.5.1     tidyverse_2.0.0  
[11] ale_0.3.0         keras_2.15.0      tensorflow_2.16.0

loaded via a namespace (and not attached):
 [1] future_1.34.0     generics_0.1.3    fortunes_1.5-4    stringi_1.8.4     lattice_0.22-6   
 [6] listenv_0.9.1     hms_1.1.3         digest_0.6.37     magrittr_2.0.3    grid_4.4.2       
[11] timechange_0.3.0  jsonlite_1.8.9    Matrix_1.7-1      whisker_0.4.1     tfruns_1.5.3     
[16] scales_1.3.0      codetools_0.2-20  cli_3.6.3         rlang_1.1.4       parallelly_1.41.0
[21] ellipsis_0.3.2    munsell_0.5.1     base64enc_0.1-3   withr_3.0.2       tools_4.4.2      
[26] parallel_4.4.2    tzdb_0.4.0        colorspace_2.1-1  zeallot_0.1.0     globals_0.16.3   
[31] reticulate_1.40.0 assertthat_0.2.1  vctrs_0.6.5       R6_2.5.1          png_0.1-8        
[36] lifecycle_1.0.4   furrr_0.3.1       pkgconfig_2.0.3   progressr_0.15.1  pillar_1.10.1    
[41] gtable_0.3.6      glue_1.8.0        Rcpp_1.0.14       tidyselect_1.2.1  rstudioapi_0.17.1
[46] farver_2.1.2      labeling_0.4.3    compiler_4.4.2   
```

Please do the same and we can compare our results.

**Comment by snvv** [comment]
Thank you
Setting parallel = 0 was required for the process to complete successfully.  Other configurations resulted in failures.
Regards

**Comment by tripartio** [comment]
That means that you had to disable parallel processing to get `ale()` to work. Please submit your session information as I requested so that I can help you debug the problem. I have my suspicions but I need to see your session information to confirm them. (Again, you MUST run all the code first before printing the session information code or else it will be incomplete and not useful.)

**Comment by tripartio** [comment]
I looked into it and I've found the source of the parallelization problem. Unfortunately, the parallel processing framework that the `ale` package uses is incompatible with the `reticulate` framework that supports Python backends like `keras` and `tensorflow`: [A Future for R: Non-Exportable Objects](https://cran.r-project.org/web/packages/future/vignettes/future-4-non-exportable-objects.html). That article explains the details. In short, every parallel worker would have to install its own version of Python just to run your `keras` code in parallel.

So, unfortunately, `ale` in R will probably never support parallelization for Python-based R packages like `keras`. :-( But at least it works in sequential code with the `parallel = 0` argument, as you've seen.

**Comment by snvv** [comment]
In reality, the ale code with keras without parallelism run extremely fast in my case. Much faster than with RF and parallel = 1.
